### PR TITLE
lockstep refactoring of `ThreadCollective`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
This pull request move the restriction out of PMacc that `ThreadCollective` must be called with the same number of worker than a supercell contains.

- remove the constructor with the N-dimensional worker index
- rename private type definitions
- add documentation
- update all call to the old constructor

# Review
Please check the changes commit wise

# Tests

- [x] game of life 
- [x] default 3D KHI

### CODE REVIEW

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 1]
```diff
@@ -32,72 +32,61 @@
...
 
-    /** execute a functor for each data point
+    /** execute the user functor for each element in the full domain
      *
-     * @tparam T_Functor type of the functor
-     * @tparam T_Args type of arguments given to the functor
+     * @tparam T_Functor type of the user functor, must have a `void operator()`
+     *                   with as many arguments as args contains
+     * @tapram  T_Args type of the arguments, each type must implement an operator
+     *                 `template<typename T, typnme R> R operator(T)`
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: typnme -> typename

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 2]
```diff
@@ -32,72 +32,61 @@
...
+    }
 
-    /** execute a functor for each data point
+    /** execute the user functor for each element in the full domain
      *
-     * @tparam T_Functor type of the functor
-     * @tparam T_Args type of arguments given to the functor
+     * @tparam T_Functor type of the user functor, must have a `void operator()`
+     *                   with as many arguments as args contains
+     * @tapram  T_Args type of the arguments, each type must implement an operator
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: one space too much before `T_Args` ;)

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 3]
```diff
@@ -108,19 +97,23 @@ class ThreadCollective
...
         >{ m_workerIdx }(
-            [&]( uint32_t const linearIdx, uint32_t const )
+            [&](
+                uint32_t const linearIdx,
+                uint32_t const
+            )
             {
-                DataSpace< dim > const relativeIdx(
-                    DataSpaceOperations< dim >::template map< FullSuperCellSize >( linearIdx ) -
+                DataSpace< dim > const pos(
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: uff, pos in what sense? cell position? in-block position? ...?

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 2 above)
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: typo: `tparam`

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 4]
```diff
@@ -32,72 +32,61 @@
...
- * @tparam T_numWorkers number of workers which are used to execute this functor
+ * @tparam T_DataDomain PMacc::SuperCellDescription, compile time data domain
+ *                      description with a CORE and GUARD
+ * @tparam T_numWorkers number of workers
  */
 template<
-    typename T_BlockDomain,
-    uint32_t T_numWorkers = math::CT::volume<typename T_BlockDomain::SuperCellSize>::type::value
+    typename T_DataDomain,
+    uint32_t T_numWorkers
```
Comment (ax3l) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: why did you remove the default? :cry: 

The usage is now unnecessary verbose in user code, if you ask me.

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 5]
```diff
@@ -32,72 +32,61 @@
...
 >
 class ThreadCollective
 {
 private:
-    /** size of the inner domain */
-    using SuperCellSize = typename T_BlockDomain::SuperCellSize;
+    // size of the CORE (in elements per dimension)
+    using CoreDomainSize = typename T_DataDomain::SuperCellSize;
+    // full size if the domain including the GUARD (in elements per dimension)
```
Comment (theZiz) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: if => of

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 1 above)
Comment (theZiz) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: Wait, I thought @psychocoderHPC does at least compile and test run every change once. :open_mouth:

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp [hunk 6]
```diff
@@ -32,72 +32,61 @@
...
-     *                spanned by T_BlockDomain
-     * @param args the result of each argument of the `operator( relativeNDimensionalIdx )` call
-     *             with the relative index inside `T_BlockDomain` based on
-     *             the origin `T_BlockDomain::OffsetOrigin` is passed to the functor
+     * @param functor user defined functor
+     * @param args arguments passed to the functor
+     *             The method `template<typename T, typnme R> R operator(T)`
+     *             is called for each argument, the result is passed to the
+     *             functor `functor::operator()`.
+     *             `T` is a N-dimensional vector of an index relative to the origin
```
Comment (theZiz) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: a**n** N-dimensional

File: src/picongpu/include/fields/FieldJ.kernel [hunk 7]
```diff
@@ -342,7 +342,14 @@ struct KernelAddCurrentToEMF
...
         auto fieldJBlock = fieldJ.shift(blockCell);
 
-        ThreadCollective<BlockArea> collective(threadIndex);
+        ThreadCollective<
+            BlockArea,
+            PMacc::math::CT::volume< typename BlockArea::SuperCellSize >::type::value
+        > collective(
+            DataSpaceOperations< BlockArea::Dim >::template map<
+                typename BlockArea::SuperCellSize
+            >( threadIndex )
```
Comment (theZiz) [inline src/picongpu/include/fields/FieldJ.kernel]: I don't get it. In the old code you just passed the `threadIndex` but now a map. However the constructor https://github.com/psychocoderHPC/picongpu/blob/cfd1322c56f4e953fa2d6b96a9e5ea97fc7a55f0/src/libPMacc/include/mappings/threads/ThreadCollective.hpp#L71 did not (?) change.
So why?

File: src/picongpu/include/fields/FieldTmp.kernel [hunk 8]
```diff
@@ -75,7 +75,10 @@ namespace picongpu
             auto cachedVal = CachedBox::create < 0, typename TmpBox::ValueType > ( BlockDescription_( ) );
             Set<typename TmpBox::ValueType > set( float_X( 0.0 ) );
 
-            ThreadCollective<BlockDescription_> collective( linearThreadIdx );
+            ThreadCollective<
+                BlockDescription_,
+                PMacc::math::CT::volume< typename BlockDescription_::SuperCellSize >::type::value
+            > collective( linearThreadIdx );
```
Comment (theZiz) [inline src/picongpu/include/fields/FieldTmp.kernel]: Here it is just `linearThreadIdx` again!
But why?

File: src/picongpu/include/particles/Particles.kernel [hunk 9]
```diff
@@ -225,7 +225,10 @@ struct KernelMoveAndMarkParticles
        auto fieldBBlock = fieldB.shift(blockCell);
 
        nvidia::functors::Assign assign;
-       ThreadCollective<BlockDescription_> collective(linearThreadIdx);
+       ThreadCollective<
+           BlockDescription_,
+           PMacc::math::CT::volume< typename BlockDescription_::SuperCellSize >::type::value
+       > collective( linearThreadIdx );
```
Comment (theZiz) [inline src/picongpu/include/particles/Particles.kernel]: Same

File: src/picongpu/include/particles/bremsstrahlung/Bremsstrahlung.tpp [hunk 10]
```diff
@@ -83,7 +83,11 @@ void Bremsstrahlung<T_IonSpecies, T_ElectronSpecies, T_PhotonSpecies>::init
     nvidia::functors::Assign assign;
     /* copy fields from global to shared */
     const auto fieldIonDensityBlock = ionDensityBox.shift(blockCell);
-    ThreadCollective<BlockArea> collective(linearThreadIdx);
+
+    ThreadCollective<
+        BlockArea,
+        PMacc::math::CT::volume< typename BlockArea::SuperCellSize >::type::value
+    > collective( linearThreadIdx );
```
Comment (theZiz) [inline src/picongpu/include/particles/bremsstrahlung/Bremsstrahlung.tpp]: Same (I will not continue writing Same...)

File: src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp [hunk 11]
```diff
@@ -155,7 +156,14 @@ struct PhotonCreator
...
         /* copy fields from global to shared */
         auto fieldBBlock = bBox.shift(blockCell);
-        ThreadCollective<BlockArea> collective(linearThreadIdx);
+        ThreadCollective<
+            BlockArea,
+            PMacc::math::CT::volume< typename BlockArea::SuperCellSize >::type::value
+        > collective(
+            DataSpaceOperations< BlockArea::Dim >::template map<
+                typename BlockArea::SuperCellSize
+            >( linearThreadIdx )
```
Comment (theZiz) [inline src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp]: Aaaand a map again. Whyever...

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 4 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: I removed that because it is mixing the data domain and the worker domain and creates a one to one relation between both. 

SHort: this is not compatible with the lock step programming model

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 1 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: @theZiz it's an documentation bug (can't find that with a test), but wait you are my CI test for documentation ^^

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 6 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: I think this must be a `a` and not an `an`

File: src/picongpu/include/fields/FieldJ.kernel (hunk 7 above)
Comment (psychocoderHPC) [inline src/picongpu/include/fields/FieldJ.kernel]: I removed the constructor for n-dimensional thread indices. Therefore we need to map the n-dimensional index to a linear index.
In the old code we used a one to one relation between the number of worker and number of elements in a supercell, this is not longer valid for lock step programming

File: src/picongpu/include/fields/FieldTmp.kernel (hunk 8 above)
Comment (psychocoderHPC) [inline src/picongpu/include/fields/FieldTmp.kernel]: There is only a constructor with a linear index available.

File: src/picongpu/include/particles/Particles.kernel (hunk 9 above)
Comment (psychocoderHPC) [inline src/picongpu/include/particles/Particles.kernel]: [see above](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2101#discussion_r121944393)

File: src/picongpu/include/particles/bremsstrahlung/Bremsstrahlung.tpp (hunk 10 above)
Comment (psychocoderHPC) [inline src/picongpu/include/particles/bremsstrahlung/Bremsstrahlung.tpp]: [see above](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2101#discussion_r121944393)

File: src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp (hunk 11 above)
Comment (psychocoderHPC) [inline src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp]: [see above](https://github.com/ComputationalRadiationPhysics/picongpu/pull/2101#discussion_r121944393)

Comment (psychocoderHPC) [inline src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp]: this is  bug, thanks for spotting

**Comment by ax3l** [comment]
@theZiz pls change your review state if the answers satisfy your requests

File: src/libPMacc/include/mappings/threads/ThreadCollective.hpp (hunk 6 above)
Comment (theZiz) [inline src/libPMacc/include/mappings/threads/ThreadCollective.hpp]: It is spoken "an EN-dimensional". Are you really saying "a Nnnnnn-dimensional"? Sounds funny. A bit Polish. They love consonants without vocals. "W" is a very often used word in Poland. Anyway, no real reason to block this PR...
